csharp/src/Apache.Arrow/Builder/BufferBuilder.cs

```diff
@@ -0,0 +1,356 @@
+
+                // Fill byte buffer
+                BitOverhead.Fill(bits.Slice(0, available));
+
+                // Commit to memory and reset
+                CommitBitBuffer();
+
+                // Fill new bit buffer
+                BitOverhead.Fill(bits.Slice(available));
+            }
+            else
+            {
+                EnsureAdditionalBytes(1);
+                // Raw add to memory
+                Memory.Span[ByteLength] = byteValue;
+                ByteLength++;
+            }
+
+            return this;
+        }
+
+        public IBufferBuilder AppendBytes(ReadOnlySpan<byte> bytes)
+        {
+            if (BitOverhead.Length == 0)
+            {
+                EnsureAdditionalBytes(bytes.Length);
+                // Raw Span copy to memory
+                bytes.CopyTo(Memory.Span.Slice(ByteLength, bytes.Length));
+                ByteLength += bytes.Length;
+            }
+            else
+            {
+                // Convert Bytes to Bits streamed in batchsize = 64
+                int offset = 0;
+                while (offset < bytes.Length)
+                {
+                    int remainingBytes = bytes.Length - offset;
+                    int bufferLength = Math.Min(128, remainingBytes);
+
+                    // Bits span
+                    Span<bool> buffer = new bool[bufferLength * 8];
+
+                    // Fill bits
+                    BitUtility.ToBits(buffer, bytes.Slice(offset, bufferLength));
+
+                    // Append batch bits
+                    AppendBits(buffer);
+                    offset += bufferLength;
+                }
+            }
+
+            return this;
+        }
+
+        public IBufferBuilder AppendStruct(bool value) => AppendBit(value);
+        public IBufferBuilder AppendStruct(byte value) => AppendByte(value);
+        public IBufferBuilder AppendStruct<T>(T value) where T : struct
+        {
+#if NETCOREAPP3_1_OR_GREATER
+            return AppendStructs(MemoryMarshal.CreateReadOnlySpan(ref value, 1));
+#else
+            return AppendStructs<T>(new T[] { value }.AsSpan());
+#endif
+        }
+
+        public IBufferBuilder AppendStructs(ReadOnlySpan<bool> values) => AppendBits(values);
+        public IBufferBuilder AppendStructs(ReadOnlySpan<byte> values) => AppendBytes(values);
+        public IBufferBuilder AppendStructs<T>(ReadOnlySpan<T> values) where T : struct
+        {
+            ReadOnlySpan<byte> bytes = MemoryMarshal.AsBytes(values);
+            return AppendBytes(bytes);
+        }
+        public IBufferBuilder AppendStructs<T>(T value, int count) where T : struct
+        {
+            Span<T> span = new T[count];
+
+            for (int i = 0; i < count; i++)
+                span[i] = value;
+
+            return AppendStructs<T>(span);
+        }
+
+        internal IBufferBuilder ReserveBytes(int numBytes)
+        {
+            EnsureAdditionalBytes(numBytes);
```

Comment:
Coming from the C++ implementation I would be surprised to learn that `ReserveBytes` reserved _additional_ bytes and didn't simply ensure bytes.  Maybe name this method `ReserveAdditionalBytes`?